### PR TITLE
Select language on first start based on user locale

### LINKED ISSUE #2459: Use system locale to set language on first execution
https://en.cppreference.com/w/cpp/locale/locale
https://en.cppreference.com/w/cpp/io/ios_base/getloc

Most games do this.

### PR DESCRIPTION
**Pull request by Robyt3**
Closes #2459.

## Checklist

- [X] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [X] Tested in combination with possibly related configuration options
- [X] Written a unit test (especially base/) or added coverage to integration test
- [X] Considered possible null pointers and out of bounds array indexing
- [X] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/base/system.cpp [hunk 1]
```diff
@@ -4266,6 +4268,104 @@ int os_version_str(char *version, int length)
...
+		"LC_COLLATE",
+		"LC_MONETARY",
+		"LC_MESSAGES",
+		"LC_PAPER",
+		"LC_NAME",
+		"LC_ADDRESS",
+		"LC_TELEPHONE",
+		"LC_MEASUREMENT",
+		"LC_IDENTIFICATION",
+		"LANGUAGE", // check last because it does not specify the region
```
Comment (heinrich5991) [inline src/base/system.cpp:4312]: This should probably only check `LC_ALL`, `LC_MESSAGES` and `LANG` in that order. Whether you use a German paper size should not have an effect on DDNet's language.

From `man 7 locale`:
> (1)  If there is a non-null environment variable LC_ALL, the value of LC_ALL is used.
>
> (2)  If an environment variable with the same name as one of the categories above exists and is non-null, its value is used for that category.
>
> (3)  If there is a non-null environment variable LANG, the value of LANG is used.

File: src/base/system.cpp [hunk 2]
```diff
@@ -4266,6 +4268,104 @@ int os_version_str(char *version, int length)
...
+		}
+	}
+
+	if(!found_env_variable)
+	{
+		// Fallback to std::locale, which does not work reliably.
+		std::string temp_locale = std::locale("").name();
+		// The string returned by this function may contain multiple locale
+		// specifications in the form "key1=value1;key2=value2;..." in which
+		// case we extract just one (arbitrary) value.
```
Comment (heinrich5991) [inline src/base/system.cpp:4333]:
```suggestion
		// case we just arbitrarily extract the first value.
```

File: src/base/system.cpp [hunk 3]
```diff
@@ -4266,6 +4268,104 @@ int os_version_str(char *version, int length)
...
+		{
+			str_copy(locale, env_value, length);
+			found_env_variable = true;
+			break;
+		}
+	}
+
+	if(!found_env_variable)
+	{
+		// Fallback to std::locale, which does not work reliably.
```
Comment (heinrich5991) [inline src/base/system.cpp:4329]: I don't think we actually need that fallback to be honest.

File: src/base/system.h [hunk 4]
```diff
@@ -2586,6 +2594,18 @@ int secure_rand_below(int below);
 */
 int os_version_str(char *version, int length);
 
+/**
+ * Returns a string of the preferred locale of the user / operating system.
+ * The string conforms to [RFC 3066](https://www.ietf.org/rfc/rfc3066.txt)
+ * and only contains the characters `a`-`z`, `A`-`Z`, `0`-`9` and `-`.
```
Comment (heinrich5991) [inline src/base/system.h:2600]: Should probably mention that it falls back to en-US if no locale was found.

**Review by heinrich5991 — COMMENTED** [review]
Thanks for working on this. I think it mostly looks good, see my comments.

For Steam, we probably want to use this instead: https://partner.steamgames.com/doc/api/ISteamApps#GetCurrentGameLanguage. I don't think implementing that is necessary to get the PR merged though.

File: src/base/system.cpp (hunk 1 above)
Comment (Robyt3) [reply src/base/system.cpp]: Done.

File: src/base/system.h (hunk 4 above)
Comment (Robyt3) [reply src/base/system.h:2600]: Done.

File: src/base/system.cpp (hunk 2 above)
Comment (Robyt3) [reply src/base/system.cpp]: Fallback was removed.

**Comment by heinrich5991** [comment]
bors r+

Thanks!

**Pull request merged**